File: maths/nombres/disarium/disarium.py
# -----------------------------------------------
# Fonction disarium
# -----------------------------------------------
# Un "disarium number" est un nombre qui est égal
# à la somme de ses chiffres élevés à la puissan-
# de leur index plus un dans le nombre.
# Ex. : 1^1 + 3^2 + 5^3 = 135
# -----------------------------------------------


# -----------------------------------------------
# Version courte
# -----------------------------------------------
# Cette version permet de comprendre le contexte
# de la fonction et sa logique.

def disarium(nombre):
    nombre_str = str(nombre)
    longueur = len(nombre_str)
    
    somme = sum(int(nombre_str[i]) ** (i + 1) for
                i in range(longueur))
        
    return somme == nombre



# -----------------------------------------------
# Application
# -----------------------------------------------

nombres = [120, 133, 135, 145]
for nb in nombres:
    res = disarium(nb)
    if res:
        print(f"{nb} est disarium.")
    else:
        print(f"{nb} n'est pas disarium.")



# -----------------------------------------------
# Version golf
# -----------------------------------------------
# Version condensee et optimisee du code, utili-
# sant des noms de variables courts et combinant
# certaines operations pour reduire la taille du
# code. Pour la beaute du geste !

D=lambda n:sum(int(str(n)[i])**(i+1)for i in 
range(len(str(n))))==n



# -----------------------------------------------
# Version detaillee
# -----------------------------------------------
# Cette version permet de suivre pas a pas l'exe-
# cution de la fonction.

def disarium_explication(nombre): #
    nombre_str = str(nombre)
    longueur = len(nombre_str)

    somme = 0

    for i in range(0, longueur):
        chiffre = int(nombre_str[i])
        puissance = i + 1
        somme += chiffre ** puissance
        
    return somme == nombre



# -----------------------------------------------
# Version commentee
# -----------------------------------------------
# Similaire a la version detaillee, mais avec des
# commentaires concis afin d'expliquer les etapes
# principales de la fonction.


def disarium_commentee(nombre): #
    # Convertir le nombre en chaîne.
    nombre_str = str(nombre)
    
    # Calculer la longueur de la chaîne.
    longueur = len(nombre_str)
    
    # Initialiser la somme des chiffres à zéro.
    somme = 0
    
    # Parcourir chaque chiffre dans la chaîne.
    for i in range(0, longueur):
        # Convertir le caractère en chiffre.
        chiffre = int(nombre_str[i])
        
        # Définir la puissance en fonction de l'index.
        puissance = i + 1
        
        # Ajouter la puissance du chiffre à la somme.
        somme += chiffre ** puissance
    
    # Retourner la solution.
    return somme == nombre









#

File: nombre/disarium/disarium.js
//-----------------------------------------------
// Fonction disarium
/*-----------------------------------------------
/ Un "disarium number" est un nombre qui est égal
/ à la somme de ses chiffres élevés à la puissan-
/ de leur index plus un dans le nombre.
/ Ex. : 1^1 + 3^2 + 5^3 = 135
/----------------------------------------------*/


//-----------------------------------------------
// Version courte
//-----------------------------------------------

function disarium(nombre) {
    
    let somme = nombre
        .toString()
        .split('')
        .reduce((acc, chiffre, i) => acc +
        Math.pow(parseInt(chiffre), i + 1), 0);

    return somme === nombre;
}



//-----------------------------------------------
// Application
//-----------------------------------------------

let nombres = [120, 133, 135, 89];
for (let nb of nombres) {
    let res = disarium(nb);
    if (res) {
        console.log(nb, "est disarium.");
    }
    else {
        console.log(nb, "n'est pas disarium.");
    }
}



//-----------------------------------------------
// Version golf
//-----------------------------------------------

/* En réflexion ;) */

let D=n=>n=== n.toString().split('').reduce((a,c,
i)=>a+Math.pow(parseInt(c),i+1),0);



//-----------------------------------------------
// Version detaillee
//-----------------------------------------------


function disarium_detaillee(nombre) {
    let nombre_str = nombre.toString();
    let longueur = nombre_str.length;
    
    let somme = 0;
    
    for (let i = 0; i < longueur; i++) {
        let chiffre = parseInt(nombre_str[i]);
        let puissance = i + 1;
        somme += Math.pow(chiffre, puissance);
    }
    
    return somme === nombre;
}



//-----------------------------------------------
// Version commentee
//-----------------------------------------------

function disarium_commentee(nombre) {
    // Convertir le nombre en chaîne.
    let nombre_str = nombre.toString();
    
    // Calculer la longueur de la chaîne.
    let longueur = nombre_str.length;
    
    // Initialiser la somme des chiffres a zéro.
    let somme = 0;
    
    // Parcourir chaque chiffre dans la chaîne.
    for (let i = 0; i < longueur; i++) {
        // Convertir le caractère en chiffre.
        let chiffre = parseInt(nombre_str[i]);
        
        // Définir la puissance en fonction de l'index.
        let puissance = i + 1;
        
        // Ajouter la puissance du chiffre à la somme.
        somme += Math.pow(chiffre, puissance);
    }
    
    // Retourner la solution
    return somme === nombre;
}









//